FDK17プロジェクト/コード/03.サウンド/SoundGroupLevelController.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Collections.Specialized;
+using FDK.ExtensionMethods;
+
+namespace FDK
+{
+    /// <summary>
+    /// SoundGroupLevelController holds the current sound level value for each
+    /// of the unique sound groups, along with an increment by which they can
+    /// easily be adjusted.
+    ///
+    /// Configuration changes to the sound group levels are provided to the
+    /// controller via binding code which allows CConfigIni and
+    /// SoundGroupLevelController to be unaware of one another.
+    /// See ConfigIniToSoundGroupLevelControllerBinder for more details.
+    ///
+    /// Dynamic adjustment of sound group levels during song selection and song
+    /// playback are managed via a small dependency taken by the respective
+    /// stage classes. See KeyboardSoundGroupLevelControlHandler and its usages
+    /// for more details.
+    ///
+    /// As new sound objects are created, including when reloading sounds due
+    /// to a changer in audio output device, SoundGroupLevelController ensures
+    /// that they are provided with the current level for their associated
+    /// sound group by subscribing to notifications regarding changes to a
+    /// collection of sound objects provided during construction. This
+    /// observable collection comes from the sound manager, but without either
+    /// it or this class being directly aware of one another.
+    ///
+    /// As sound group levels are changed, SoundGroupLevelController updates
+    /// all existing sound objects group levels by iterating that same
+    /// observable collection.
+    /// </summary>
+    public sealed class SoundGroupLevelController
```

Comment:
Here at long last is the new SoundGroupLevelController. Please read its doc comment just above.